docs/architecture/adr-037-peer-behaviour.md

```diff
@@ -0,0 +1,146 @@
+coupling `p2p.Switch`.  Introduce a ErrPeer to provide 
+concrete reasons for stopping peers.
+
+### Implementation Changes
+
+PeerBehaviour then becomes an interface for signaling peer errors as well
+as marking peers for `good`.
+
+XXX: It might be better to pass p2p.ID instead of the whole peer but as
+a first draft maintain the underlying implementation as much as
+possible.
+
+```go
+type PeerBehaviour interface {
+    Errored(peer Peer, reason ErrPeer)
+    MarkPeerAsGood(peer Peer)
+}
+```
+
+Instead of signaling peers to stop with arbitrary reasons:
+`reason interface{}` 
+
+We introduce a concrete error type ErrPeer:
+```go
+type ErrPeer int
+
+const (
+    ErrPeerUnknown = iota
+    ErrPeerBadMessage
+    ErrPeerMessageOutofOrder
+    ...
+)
+```
+
+As a first iteration we provide a concrete implementation which wraps
+the switch:
+```go
+type SwitchedPeerBehaviour struct {
+    sw *Switch
+}
+
+func (spb *SwitchedPeerBehaviour) Errored(peer Peer, reason ErrPeer) {
+    spb.sw.StopPeerForError(peer, reason)
+}
+
+func (spb *SwitchedPeerBehaviour) MarkPeerAsGood(peer Peer) {
+    spb.sw.MarkPeerAsGood(peer)
+}
+
+func NewSwitchedPeerBehaviour(sw *Switch) *SwitchedPeerBehaviour {
+    return &SwitchedPeerBehaviour{
+        sw: sw,
+    }
+}
+```
+
+Reactors, which are often difficult to unit test (Cf. [#3506](https://github.com/tendermint/tendermint/pull/3506)). could use an implementation which exposes the signals produced by the reactor in
+manufactured scenarios:
+
+```go
+type PeerErrors map[Peer][]ErrPeer
+type GoodPeers map[Peer]bool
+
+type StorePeerBehaviour struct {
+    pe PeerErrors
+    gp GoodPeers
+}
+
+func NewStorePeerBehaviour() *StorePeerBehaviour{
+    return &StorePeerBehaviour{
+        pe: make(PeerErrors),
+        gp: GoodPeers{},
+    }
+}
+
+func (spb StorePeerBehaviour) Errored(peer Peer, reason ErrPeer) {
+    if _, ok := spb.pe[peer]; !ok {
+        spb.pe[peer] = []ErrPeer{reason}
+    } else {
+        spb.pe[peer] = append(spb.pe[peer], reason)
+    }
+}
+
+func (mpb *StorePeerBehaviour) GetPeerErrors() PeerErrors {
+    return mpb.pe
+}
+
+func (spb *StorePeerBehaviour) MarkPeerAsGood(peer Peer) {
+    if _, ok := spb.gp[peer]; !ok {
+        spb.gp[peer] = true
+    }
+}
+
+func (spb *StorePeerBehaviour) GetGoodPeers() GoodPeers {
+    return spb.gp
+}
+```
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+    * De-couple signaling from acting upon peer behaviour.
+    * Reduce the coupling of reactors and the Switch and the network
+      stack
+    * The responsibility of managing peer behaviour can be migrated to
+      a single component instead of split between the switch and the
+      address book.
+
+### Negative
+
+    * The first iteration will simply wrap the Switch and introduce a
+      level of indirection.
+
+### Neutral
+
+## References
+
+* [#2067](https://github.com/tendermint/tendermint/issues/2067): P2P Refactor
```

Comment:
Nit: you can use the footnote style for linking (prob. there is an actual name for this) these to increase readability in the text above and to re-use the link. 
E.g.
```
Mention [#2067] in the text like this and use it again somewhere else, 
e.g. here: [#2067] and link to it once at the end (or somewhere):

[#2067]: https://github.com/tendermint/tendermint/issues/2067
```

Reply:
Do you mean like this?
```suggestion
* [#2067]: https://github.com/tendermint/tendermint/issues/2067
```
or possibly use  [the markdown footnote](https://www.markdownguide.org/extended-syntax/#footnotes) to link the text to the footnote to the actual link?

Reply:
> or possibly use the markdown footnote to link the text to the footnote to the actual link? 

This one. I think it makes the document slightly easier to read and you can re-use the reference in the text.

Reply:
Maybe I'm missing something, but Github markdown doesn't appear to support that particular `[^#]` features but we can accomplish something similar by linking to the Reference section with sub text. 

![4d7](https://user-images.githubusercontent.com/246/55905979-863ecb00-5bd3-11e9-9b8b-8fa2c9a8da5f.png)

Reply:
Haha, where did that awesome meme go! I really just meant to replace the 2 (or more) occurrences of links to issue with what is described in https://github.com/tendermint/tendermint/pull/3539#discussion_r273869696
(and for above reasons)
github supports this.
But NVM. The PR is really good as it is and it was just a tiny nit.